Libraries/Utilities/utf8.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule utf8
+ * @flow
+ */
+'use strict';
+
+class ByteVector {
+  _storage: Uint8Array;
+  _sizeWritten: number;
+
+  constructor(size) {
+    this._storage = new Uint8Array(size);
+    this._sizeWritten = 0;
+  }
+
+  push(value: number): ByteVector {
+    const i = this._sizeWritten;
+    if (i === this._storage.length) {
+      this._realloc();
+    }
+    this._storage[i] = value;
+    this._sizeWritten = i + 1;
+    return this;
+  }
+
+  getBuffer(): ArrayBuffer {
+    return this._storage.buffer.slice(0, this._sizeWritten);
+  }
+
+  _realloc() {
+    const storage = this._storage;
+    this._storage = new Uint8Array(align(storage.length * 1.5));
+    this._storage.set(storage);
+  }
+}
+
+/*eslint-disable no-bitwise */
+exports.encode = (string: string): ArrayBuffer => {
```

Comment:
I think the complexity of crypto and character encodings is very different. I have been looking for a compact utf-8 encoding lib, but didn’t find anything that was compact *and* solid at the same time. My implementation adds ~1KB when minified.

Do you have any specific concerns? We wouldn’t need to do this if we could pass the binary response data from native to JS wrapped into a host object we’d implement. That host object could pass of character encoding to a system service. I have been looking into JSC’s API for that, but as long as we pass messages as JSON, we can’t do that.